File: apps/api/app/routers/v1/sso.py
"""
SSO/SAML API endpoints for enterprise authentication
"""

from fastapi import APIRouter, Depends, HTTPException, Request, Response, status
from fastapi.responses import RedirectResponse
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from typing import Optional, Dict, Any, List
from pydantic import BaseModel, Field
import logging

from app.database import get_db
from app.dependencies import get_current_user, require_admin
from ...models import User, Organization, SSOConfiguration, SSOProvider, SSOStatus
from app.services.sso_service import SSOService

logger = logging.getLogger(__name__)

router = APIRouter(
    prefix="/sso",
    tags=["SSO/SAML"],
    responses={404: {"description": "Not found"}},
)


# Pydantic models for request/response
class SSOConfigurationCreate(BaseModel):
    """Create SSO configuration request"""
    provider: SSOProvider
    # SAML fields
    saml_metadata_url: Optional[str] = None
    saml_metadata_xml: Optional[str] = None
    saml_sso_url: Optional[str] = None
    saml_slo_url: Optional[str] = None
    saml_certificate: Optional[str] = None
    saml_entity_id: Optional[str] = None
    
    # OIDC fields
    oidc_issuer: Optional[str] = None
    oidc_client_id: Optional[str] = None
    oidc_client_secret: Optional[str] = None
    oidc_discovery_url: Optional[str] = None
    
    # Common settings
    jit_provisioning: bool = True
    default_role: str = "member"
    attribute_mapping: Dict[str, str] = Field(default_factory=dict)
    allowed_domains: List[str] = Field(default_factory=list)


class SSOConfigurationUpdate(BaseModel):
    """Update SSO configuration request"""
    enabled: Optional[bool] = None
    # SAML fields
    saml_metadata_url: Optional[str] = None
    saml_metadata_xml: Optional[str] = None
    saml_sso_url: Optional[str] = None
    saml_slo_url: Optional[str] = None
    saml_certificate: Optional[str] = None
    
    # OIDC fields
    oidc_issuer: Optional[str] = None
    oidc_client_id: Optional[str] = None
    oidc_client_secret: Optional[str] = None
    oidc_discovery_url: Optional[str] = None
    
    # Settings
    jit_provisioning: Optional[bool] = None
    default_role: Optional[str] = None
    attribute_mapping: Optional[Dict[str, str]] = None
    allowed_domains: Optional[List[str]] = None


class SSOConfigurationResponse(BaseModel):
    """SSO configuration response"""
    id: str
    organization_id: str
    provider: SSOProvider
    status: SSOStatus
    enabled: bool
    
    # SAML info (sanitized)
    saml_entity_id: Optional[str] = None
    saml_acs_url: Optional[str] = None
    saml_sso_url: Optional[str] = None
    
    # OIDC info (sanitized)
    oidc_issuer: Optional[str] = None
    oidc_client_id: Optional[str] = None
    oidc_authorization_url: Optional[str] = None
    
    # Settings
    jit_provisioning: bool
    default_role: str
    allowed_domains: List[str]
    
    created_at: str
    updated_at: str


class SSOInitiateRequest(BaseModel):
    """SSO login initiation request"""
    organization_slug: str
    return_url: Optional[str] = None


class SSOTestRequest(BaseModel):
    """SSO configuration test request"""
    configuration_id: str


# SSO service dependency
def get_sso_service() -> SSOService:
    """
    Create SSO service instance with dependencies.
    This is a placeholder - in production, inject actual db, cache, and jwt_service
    """
    # TODO: Replace with actual dependency injection
    return None  # Will be properly initialized when dependencies are available


@router.post("/configurations", response_model=SSOConfigurationResponse)
async def create_sso_configuration(
    organization_id: str,
    config: SSOConfigurationCreate,
    current_user: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db)
):
    """
    Create SSO configuration for an organization
    
    Requires admin privileges for the organization.
    """
    try:
        # Verify organization exists and user is admin
        org = await db.get(Organization, organization_id)
        if not org:
            raise HTTPException(status_code=404, detail="Organization not found")
        
        # Check if SSO config already exists
        result = await db.execute(select(SSOConfiguration).where(
            SSOConfiguration.organization_id == organization_id
        ))
        existing = result.scalar_one_or_none()
        if existing:
            raise HTTPException(
                status_code=400, 
                detail="SSO configuration already exists for this organization"
            )
        
        # Create configuration based on provider
        if config.provider == SSOProvider.SAML:
            result = await sso_service.create_saml_configuration(
                organization_id=organization_id,
                metadata_url=config.saml_metadata_url,
                metadata_xml=config.saml_metadata_xml,
                sso_url=config.saml_sso_url,
                slo_url=config.saml_slo_url,
                certificate=config.saml_certificate,
                entity_id=config.saml_entity_id,
                jit_provisioning=config.jit_provisioning,
                default_role=config.default_role,
                attribute_mapping=config.attribute_mapping,
                allowed_domains=config.allowed_domains
            )
        elif config.provider == SSOProvider.OIDC:
            result = await sso_service.create_oidc_configuration(
                organization_id=organization_id,
                issuer=config.oidc_issuer,
                client_id=config.oidc_client_id,
                client_secret=config.oidc_client_secret,
                discovery_url=config.oidc_discovery_url,
                jit_provisioning=config.jit_provisioning,
                default_role=config.default_role,
                attribute_mapping=config.attribute_mapping,
                allowed_domains=config.allowed_domains
            )
        else:
            raise HTTPException(
                status_code=400,
                detail=f"Provider {config.provider} not yet supported"
            )
        
        return SSOConfigurationResponse(**result)
        
    except Exception as e:
        logger.error(f"Failed to create SSO configuration: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/configurations/{organization_id}", response_model=SSOConfigurationResponse)
async def get_sso_configuration(
    organization_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    Get SSO configuration for an organization
    
    Returns the SSO configuration if the user belongs to the organization.
    """
    try:
        config = await sso_service.get_configuration(organization_id)
        if not config:
            raise HTTPException(status_code=404, detail="SSO configuration not found")
        
        # Verify user belongs to organization
        # TODO: Add proper organization membership check
        
        return SSOConfigurationResponse(**config)
        
    except Exception as e:
        logger.error(f"Failed to get SSO configuration: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/configurations/{organization_id}", response_model=SSOConfigurationResponse)
async def update_sso_configuration(
    organization_id: str,
    update: SSOConfigurationUpdate,
    current_user: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db)
):
    """
    Update SSO configuration for an organization
    
    Requires admin privileges for the organization.
    """
    try:
        result = await sso_service.update_configuration(
            organization_id, 
            update.dict(exclude_unset=True)
        )
        if not result:
            raise HTTPException(status_code=404, detail="SSO configuration not found")
        
        return SSOConfigurationResponse(**result)
        
    except Exception as e:
        logger.error(f"Failed to update SSO configuration: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/configurations/{organization_id}")
async def delete_sso_configuration(
    organization_id: str,
    current_user: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db)
):
    """
    Delete SSO configuration for an organization
    
    Requires admin privileges for the organization.
    """
    try:
        result = await sso_service.delete_configuration(organization_id)
        if not result:
            raise HTTPException(status_code=404, detail="SSO configuration not found")
        
        return {"message": "SSO configuration deleted successfully"}
        
    except Exception as e:
        logger.error(f"Failed to delete SSO configuration: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/initiate")
async def initiate_sso(
    request: SSOInitiateRequest,
    db: AsyncSession = Depends(get_db)
):
    """
    Initiate SSO login flow
    
    Returns redirect URL to identity provider.
    """
    try:
        # Get organization by slug
        org_result = await db.execute(select(Organization).where(
            Organization.slug == request.organization_slug
        ))
        org = org_result.scalar_one_or_none()
        if not org:
            raise HTTPException(status_code=404, detail="Organization not found")
        
        # Get SSO configuration
        config = await sso_service.get_configuration(str(org.id))
        if not config or not config.get("enabled"):
            raise HTTPException(
                status_code=400, 
                detail="SSO is not enabled for this organization"
            )
        
        # Initiate based on provider
        if config["provider"] == SSOProvider.SAML.value:
            result = await sso_service.initiate_saml_sso(
                str(org.id), 
                request.return_url
            )
        elif config["provider"] == SSOProvider.OIDC.value:
            result = await sso_service.initiate_oidc_sso(
                str(org.id), 
                request.return_url
            )
        else:
            raise HTTPException(
                status_code=400,
                detail=f"Provider {config['provider']} not supported"
            )
        
        return RedirectResponse(url=result["redirect_url"], status_code=302)
        
    except Exception as e:
        logger.error(f"Failed to initiate SSO: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/saml/acs")
async def saml_acs(
    request: Request,
    db: AsyncSession = Depends(get_db)
):
    """
    SAML Assertion Consumer Service (ACS) endpoint
    
    Handles SAML response from identity provider.
    """
    try:
        form_data = await request.form()
        saml_response = form_data.get("SAMLResponse")
        relay_state = form_data.get("RelayState", "")
        
        if not saml_response:
            raise HTTPException(status_code=400, detail="Missing SAML response")
        
        # Handle SAML response
        result = await sso_service.handle_saml_response(saml_response, relay_state)
        
        # Create session and redirect
        # TODO: Create JWT tokens and session
        access_token = result.get("access_token")
        refresh_token = result.get("refresh_token")
        return_url = result.get("return_url", "/")
        
        # Set cookies and redirect
        response = RedirectResponse(url=return_url, status_code=302)
        response.set_cookie(
            key="access_token",
            value=access_token,
            httponly=True,
            secure=True,
            samesite="lax"
        )
        response.set_cookie(
            key="refresh_token",
            value=refresh_token,
            httponly=True,
            secure=True,
            samesite="lax"
        )
        
        return response
        
    except Exception as e:
        logger.error(f"SAML ACS error: {str(e)}")
        raise HTTPException(status_code=500, detail="SSO authentication failed")


@router.post("/saml/slo")
async def saml_slo(
    request: Request,
    db: AsyncSession = Depends(get_db)
):
    """
    SAML Single Logout (SLO) endpoint
    
    Handles logout request from identity provider.
    """
    try:
        form_data = await request.form()
        saml_request = form_data.get("SAMLRequest")
        saml_response = form_data.get("SAMLResponse")
        relay_state = form_data.get("RelayState", "")
        
        if saml_request:
            # Handle logout request from IdP
            result = await sso_service.handle_saml_logout_request(
                saml_request, 
                relay_state
            )
        elif saml_response:
            # Handle logout response from IdP
            result = await sso_service.handle_saml_logout_response(
                saml_response, 
                relay_state
            )
        else:
            raise HTTPException(
                status_code=400, 
                detail="Missing SAML request or response"
            )
        
        # Clear session and redirect
        response = RedirectResponse(url=result.get("redirect_url", "/"), status_code=302)
        response.delete_cookie("access_token")
        response.delete_cookie("refresh_token")
        
        return response
        
    except Exception as e:
        logger.error(f"SAML SLO error: {str(e)}")
        raise HTTPException(status_code=500, detail="Logout failed")


@router.get("/oidc/callback")
async def oidc_callback(
    code: str,
    state: str,
    error: Optional[str] = None,
    error_description: Optional[str] = None,
    db: AsyncSession = Depends(get_db)
):
    """
    OIDC callback endpoint
    
    Handles OAuth2/OIDC response from identity provider.
    """
    try:
        if error:
            logger.error(f"OIDC error: {error} - {error_description}")
            raise HTTPException(
                status_code=400, 
                detail=error_description or "Authentication failed"
            )
        
        # Handle OIDC callback
        result = await sso_service.handle_oidc_callback(code, state)
        
        # Create session and redirect
        access_token = result.get("access_token")
        refresh_token = result.get("refresh_token")
        return_url = result.get("return_url", "/")
        
        # Set cookies and redirect
        response = RedirectResponse(url=return_url, status_code=302)
        response.set_cookie(
            key="access_token",
            value=access_token,
            httponly=True,
            secure=True,
            samesite="lax"
        )
        response.set_cookie(
            key="refresh_token",
            value=refresh_token,
            httponly=True,
            secure=True,
            samesite="lax"
        )
        
        return response
        
    except Exception as e:
        logger.error(f"OIDC callback error: {str(e)}")
        raise HTTPException(status_code=500, detail="SSO authentication failed")


@router.post("/test")
async def test_sso_configuration(
    test_request: SSOTestRequest,
    current_user: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db)
):
    """
    Test SSO configuration
    
    Validates the SSO configuration without initiating a full login flow.
    """
    try:
        result = await sso_service.test_configuration(test_request.configuration_id)
        return result
        
    except Exception as e:
        logger.error(f"SSO test failed: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/metadata/{organization_id}")
async def get_sp_metadata(
    organization_id: str,
    db: AsyncSession = Depends(get_db)
):
    """
    Get Service Provider (SP) metadata for SAML configuration
    
    Returns the SP metadata XML for configuring the identity provider.
    """
    try:
        metadata = await sso_service.generate_sp_metadata(organization_id)
        if not metadata:
            raise HTTPException(status_code=404, detail="Organization not found")
        
        return Response(
            content=metadata,
            media_type="application/xml",
            headers={
                "Content-Disposition": f"attachment; filename=sp-metadata-{organization_id}.xml"
            }
        )
        
    except Exception as e:
        logger.error(f"Failed to generate SP metadata: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))

File: packages/typescript-sdk/src/sso.ts
/**
 * SSO/SAML module for the Janua TypeScript SDK
 *
 * Provides enterprise Single Sign-On capabilities with SAML 2.0 and OIDC support.
 */

import type { HttpClient } from './http-client';
import type { UUID, ISODateString } from './types';
import { ValidationError } from './errors';
import { ValidationUtils } from './utils';

// SSO Types

export enum SSOProvider {
  SAML = 'saml',
  OIDC = 'oidc',
  GOOGLE_WORKSPACE = 'google_workspace',
  AZURE_AD = 'azure_ad',
  OKTA = 'okta'
}

export enum SSOStatus {
  ACTIVE = 'active',
  PENDING = 'pending',
  DISABLED = 'disabled',
  ERROR = 'error'
}

export interface SSOConfigurationCreate {
  provider: SSOProvider;
  // SAML fields
  saml_metadata_url?: string;
  saml_metadata_xml?: string;
  saml_sso_url?: string;
  saml_slo_url?: string;
  saml_certificate?: string;
  saml_entity_id?: string;
  // OIDC fields
  oidc_issuer?: string;
  oidc_client_id?: string;
  oidc_client_secret?: string;
  oidc_discovery_url?: string;
  // Common settings
  jit_provisioning?: boolean;
  default_role?: string;
  attribute_mapping?: Record<string, string>;
  allowed_domains?: string[];
}

export interface SSOConfigurationUpdate {
  enabled?: boolean;
  // SAML fields
  saml_metadata_url?: string;
  saml_metadata_xml?: string;
  saml_sso_url?: string;
  saml_slo_url?: string;
  saml_certificate?: string;
  // OIDC fields
  oidc_issuer?: string;
  oidc_client_id?: string;
  oidc_client_secret?: string;
  oidc_discovery_url?: string;
  // Settings
  jit_provisioning?: boolean;
  default_role?: string;
  attribute_mapping?: Record<string, string>;
  allowed_domains?: string[];
}

export interface SSOConfiguration {
  id: UUID;
  organization_id: UUID;
  provider: SSOProvider;
  status: SSOStatus;
  enabled: boolean;
  // SAML info (sanitized)
  saml_entity_id?: string;
  saml_acs_url?: string;
  saml_sso_url?: string;
  // OIDC info (sanitized)
  oidc_issuer?: string;
  oidc_client_id?: string;
  oidc_authorization_url?: string;
  // Settings
  jit_provisioning: boolean;
  default_role: string;
  allowed_domains: string[];
  created_at: ISODateString;
  updated_at: ISODateString;
}

export interface SSOInitiateRequest {
  organization_slug: string;
  return_url?: string;
}

export interface SSOInitiateResponse {
  redirect_url: string;
  state?: string;
}

export interface SSOTestRequest {
  configuration_id: UUID;
}

export interface SSOTestResponse {
  success: boolean;
  provider: SSOProvider;
  metadata_valid?: boolean;
  connection_status?: 'success' | 'failed' | 'timeout';
  error_message?: string;
  details?: Record<string, any>;
}

/**
 * SSO/SAML management operations
 */
export class SSO {
  constructor(private http: HttpClient) {}

  /**
   * Create SSO configuration for an organization
   *
   * @param organizationId - Organization ID
   * @param config - SSO configuration details
   * @returns Created SSO configuration
   */
  async createConfiguration(
    organizationId: UUID,
    config: SSOConfigurationCreate
  ): Promise<SSOConfiguration> {
    // Validate organization ID
    if (!ValidationUtils.isValidUuid(organizationId)) {
      throw new ValidationError('Invalid organization ID format');
    }

    // Validate provider
    if (!Object.values(SSOProvider).includes(config.provider)) {
      throw new ValidationError('Invalid SSO provider');
    }

    // Validate SAML configuration if provider is SAML
    if (config.provider === SSOProvider.SAML) {
      if (!config.saml_metadata_url && !config.saml_metadata_xml && !config.saml_sso_url) {
        throw new ValidationError(
          'SAML configuration requires either metadata URL, metadata XML, or SSO URL'
        );
      }

      if (config.saml_metadata_url && !ValidationUtils.isValidUrl(config.saml_metadata_url)) {
        throw new ValidationError('Invalid SAML metadata URL format');
      }

      if (config.saml_sso_url && !ValidationUtils.isValidUrl(config.saml_sso_url)) {
        throw new ValidationError('Invalid SAML SSO URL format');
      }

      if (config.saml_slo_url && !ValidationUtils.isValidUrl(config.saml_slo_url)) {
        throw new ValidationError('Invalid SAML SLO URL format');
      }
    }

    // Validate OIDC configuration if provider is OIDC
    if (config.provider === SSOProvider.OIDC) {
      if (!config.oidc_issuer || !config.oidc_client_id || !config.oidc_client_secret) {
        throw new ValidationError(
          'OIDC configuration requires issuer, client ID, and client secret'
        );
      }

      if (!ValidationUtils.isValidUrl(config.oidc_issuer)) {
        throw new ValidationError('Invalid OIDC issuer URL format');
      }

      if (config.oidc_discovery_url && !ValidationUtils.isValidUrl(config.oidc_discovery_url)) {
        throw new ValidationError('Invalid OIDC discovery URL format');
      }
    }

    // Validate allowed domains
    if (config.allowed_domains && config.allowed_domains.length > 0) {
      for (const domain of config.allowed_domains) {
        if (!/^[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.[a-z]{2,}$/i.test(domain)) {
          throw new ValidationError(`Invalid domain format: ${domain}`);
        }
      }
    }

    const response = await this.http.post<SSOConfiguration>(
      `/api/v1/sso/configurations?organization_id=${organizationId}`,
      config
    );
    return response.data;
  }

  /**
   * Get SSO configuration for an organization
   *
   * @param organizationId - Organization ID
   * @returns SSO configuration if exists
   */
  async getConfiguration(organizationId: UUID): Promise<SSOConfiguration> {
    if (!ValidationUtils.isValidUuid(organizationId)) {
      throw new ValidationError('Invalid organization ID format');
    }

    const response = await this.http.get<SSOConfiguration>(
      `/api/v1/sso/configurations/${organizationId}`
    );
    return response.data;
  }

  /**
   * Update SSO configuration for an organization
   *
   * @param organizationId - Organization ID
   * @param update - Fields to update
   * @returns Updated SSO configuration
   */
  async updateConfiguration(
    organizationId: UUID,
    update: SSOConfigurationUpdate
  ): Promise<SSOConfiguration> {
    if (!ValidationUtils.isValidUuid(organizationId)) {
      throw new ValidationError('Invalid organization ID format');
    }

    // Validate URLs if provided
    if (update.saml_metadata_url && !ValidationUtils.isValidUrl(update.saml_metadata_url)) {
      throw new ValidationError('Invalid SAML metadata URL format');
    }

    if (update.saml_sso_url && !ValidationUtils.isValidUrl(update.saml_sso_url)) {
      throw new ValidationError('Invalid SAML SSO URL format');
    }

    if (update.saml_slo_url && !ValidationUtils.isValidUrl(update.saml_slo_url)) {
      throw new ValidationError('Invalid SAML SLO URL format');
    }

    if (update.oidc_issuer && !ValidationUtils.isValidUrl(update.oidc_issuer)) {
      throw new ValidationError('Invalid OIDC issuer URL format');
    }

    if (update.oidc_discovery_url && !ValidationUtils.isValidUrl(update.oidc_discovery_url)) {
      throw new ValidationError('Invalid OIDC discovery URL format');
    }

    // Validate allowed domains
    if (update.allowed_domains && update.allowed_domains.length > 0) {
      for (const domain of update.allowed_domains) {
        if (!/^[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.[a-z]{2,}$/i.test(domain)) {
          throw new ValidationError(`Invalid domain format: ${domain}`);
        }
      }
    }

    const response = await this.http.put<SSOConfiguration>(
      `/api/v1/sso/configurations/${organizationId}`,
      update
    );
    return response.data;
  }

  /**
   * Delete SSO configuration for an organization
   *
   * @param organizationId - Organization ID
   * @returns Success message
   */
  async deleteConfiguration(organizationId: UUID): Promise<{ message: string }> {
    if (!ValidationUtils.isValidUuid(organizationId)) {
      throw new ValidationError('Invalid organization ID format');
    }

    const response = await this.http.delete<{ message: string }>(
      `/api/v1/sso/configurations/${organizationId}`
    );
    return response.data;
  }

  /**
   * Initiate SSO login flow
   *
   * @param request - Organization slug and optional return URL
   * @returns Redirect URL to identity provider
   */
  async initiateSSO(request: SSOInitiateRequest): Promise<SSOInitiateResponse> {
    if (!request.organization_slug || request.organization_slug.trim().length === 0) {
      throw new ValidationError('Organization slug is required');
    }

    if (!ValidationUtils.isValidSlug(request.organization_slug)) {
      throw new ValidationError('Invalid organization slug format');
    }

    if (request.return_url && !ValidationUtils.isValidUrl(request.return_url)) {
      throw new ValidationError('Invalid return URL format');
    }

    const response = await this.http.post<SSOInitiateResponse>(
      '/api/v1/sso/initiate',
      request
    );
    return response.data;
  }

  /**
   * Test SSO configuration
   *
   * Validates the SSO configuration without initiating a full login flow.
   * Useful for troubleshooting and verification.
   *
   * @param configurationId - SSO configuration ID
   * @returns Test results with connection status
   */
  async testConfiguration(configurationId: UUID): Promise<SSOTestResponse> {
    if (!ValidationUtils.isValidUuid(configurationId)) {
      throw new ValidationError('Invalid configuration ID format');
    }

    const response = await this.http.post<SSOTestResponse>(
      '/api/v1/sso/test',
      { configuration_id: configurationId }
    );
    return response.data;
  }

  /**
   * Generate Service Provider metadata XML for SAML configuration
   *
   * Returns the SP metadata XML that should be uploaded to the identity provider.
   *
   * @param organizationId - Organization ID
   * @returns SP metadata XML string
   */
  async generateSPMetadata(organizationId: UUID): Promise<string> {
    if (!ValidationUtils.isValidUuid(organizationId)) {
      throw new ValidationError('Invalid organization ID format');
    }

    const response = await this.http.get<string>(
      `/api/v1/sso/metadata/${organizationId}`
    );
    return response.data;
  }

  /**
   * Enable SSO configuration
   *
   * @param organizationId - Organization ID
   * @returns Updated SSO configuration
   */
  async enableConfiguration(organizationId: UUID): Promise<SSOConfiguration> {
    return this.updateConfiguration(organizationId, { enabled: true });
  }

  /**
   * Disable SSO configuration
   *
   * @param organizationId - Organization ID
   * @returns Updated SSO configuration
   */
  async disableConfiguration(organizationId: UUID): Promise<SSOConfiguration> {
    return this.updateConfiguration(organizationId, { enabled: false });
  }
}
